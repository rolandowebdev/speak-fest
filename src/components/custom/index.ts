export * from './button'
export * from './header'
