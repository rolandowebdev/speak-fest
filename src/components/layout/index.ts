export * from './footer'
export * from './navigation'
export * from './page-container'
