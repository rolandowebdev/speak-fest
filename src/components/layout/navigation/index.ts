export * from './items'
export * from './link'
export * from './navigation'
export * from './link-skeleton'
