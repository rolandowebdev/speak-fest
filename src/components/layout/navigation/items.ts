import React from 'react'
import {
	User,
	PenLine,
	Dice6,
	ClipboardPen,
	MessagesSquare,
	LogIn,
} from 'lucide-react'

export type MenuLink = {
	href: string
	name: string
	visible: 'all' | 'no-auth' | 'auth'
	Icon: React.ElementType
}

export const menuLinkItems: MenuLink[] = [
	{
		href: '/',
		name: 'Threads',
		visible: 'all',
		Icon: MessagesSquare,
	},
	{
		href: '/leaderboards',
		name: 'Leaderboards',
		visible: 'all',
		Icon: Dice6,
	},
	{
		href: '/login',
		name: 'Log In',
		visible: 'no-auth',
		Icon: LogIn,
	},
	{
		href: '/register',
		name: 'Register',
		visible: 'no-auth',
		Icon: ClipboardPen,
	},
	{
		href: '/create-thread',
		name: 'New Thread',
		visible: 'auth',
		Icon: PenLine,
	},
	{
		href: '/profile',
		name: 'My Profile',
		visible: 'auth',
		Icon: User,
	},
]
