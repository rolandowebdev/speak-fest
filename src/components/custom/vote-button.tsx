import * as React from 'react'
import { Button } from '@/components/ui'
import { ThumbsDown, ThumbsUp } from 'lucide-react'

export type VoteButtonProps = {
  voteType: 'up-vote' | 'down-vote'
  isVoted: boolean
  voteCount: number
  onClick: () => void
  authStatus: 'authenticated' | 'unauthenticated' | 'loading' | undefined
}

export function VoteButton({
  voteType,
  isVoted,
  voteCount,
  onClick,
  authStatus,
}: VoteButtonProps) {
  let icon

  if (voteType === 'up-vote') {
    icon = isVoted ? (
      <ThumbsUp
        size={18}
        className="fill-blue-500 dark:fill-blue-600"
        data-testid="thumbs-up-icon"
      />
    ) : (
      <ThumbsUp size={18} data-testid="thumbs-up-icon" />
    )
  } else {
    icon = isVoted ? (
      <ThumbsDown
        size={18}
        className="fill-blue-500 dark:fill-blue-600"
        data-testid="thumbs-down-icon"
      />
    ) : (
      <ThumbsDown size={18} data-testid="thumbs-down-icon" />
    )
  }

  return (
    <Button
      variant="ghost"
      onClick={onClick}
      className="flex items-center gap-1 text-lg"
      disabled={authStatus === 'unauthenticated' || authStatus === 'loading'}>
      {icon}
      <span className="text-lg">{voteCount || 0}</span>
    </Button>
  )
}
