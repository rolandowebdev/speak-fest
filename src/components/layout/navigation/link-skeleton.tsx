import { Skeleton } from '@/components/ui'

export const NavigationLinkSkeleton = () => {
	return (
		<>
			<Skeleton className='flex items-center space-x-2 border-2 border-transparent bg-slate-200 bg-opacity-25 h-11 px-3 py-2 md:min-w-[10rem] font-medium tracking-wide transition-colors rounded-none' />
			<Skeleton className='flex items-center space-x-2 border-2 border-transparent bg-slate-200 bg-opacity-25 h-11 px-3 py-2 md:min-w-[10rem] font-medium tracking-wide transition-colors rounded-none' />
			<Skeleton className='flex items-center space-x-2 border-2 border-transparent bg-slate-200 bg-opacity-25 h-11 px-3 py-2 md:min-w-[10rem] font-medium tracking-wide transition-colors rounded-none' />
		</>
	)
}
