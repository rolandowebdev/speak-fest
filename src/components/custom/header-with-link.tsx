import * as React from 'react'
import { Button, Heading } from '@/components/ui'
import { Undo2 } from 'lucide-react'
import { useRouter } from 'next/navigation'
import { Header } from '@/components/layout'

type HeaderWithLinkProps = {
  icon: React.ReactNode
  title: string
}

export function HeaderWithLink({ icon, title }: HeaderWithLinkProps) {
  const { push } = useRouter()

  return (
    <Header>
      <Button
        data-testid="back-to-home"
        variant="link"
        className="flex items-center gap-1 px-0 text-lg text-primary"
        onClick={() => push('/')}>
        <Undo2 size={18} />
        Back to home
      </Button>
      <Heading className="flex flex-wrap items-center gap-2">
        {icon} {title}
      </Heading>
    </Header>
  )
}
