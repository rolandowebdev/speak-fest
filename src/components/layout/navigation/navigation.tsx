'use client'

import React from 'react'
import { usePathname } from 'next/navigation'
import { ChevronsLeft, ChevronsRight, PanelTopOpen } from 'lucide-react'
import { cn } from '@/libs/utils'
import { useNavigationState } from '@/hooks'
import {
	Button,
	Heading,
	Sheet,
	SheetContent,
	SheetTrigger,
	ImageBlur,
} from '@/components/ui'
import {
	NavigationLink,
	MenuLink,
	menuLinkItems,
	NavigationLinkSkeleton,
} from '@/components/layout'
import { useSession } from 'next-auth/react'

const MobileNavigation = ({
	pathname,
	isOpen,
	setIsOpen,
	filterLinkItems,
}: {
	pathname: string
	isOpen: boolean
	setIsOpen: (open: boolean) => void
	filterLinkItems: () => MenuLink[]
}) => {
	return (
		<>
			<Sheet open={isOpen} onOpenChange={setIsOpen}>
				<SheetTrigger asChild>
					<Button
						variant='outline'
						size='icon'
						className='fixed bottom-0 right-0 z-20 m-2 inline-flex shadow-sm sm:hidden'>
						<PanelTopOpen className='h-[1.2rem] w-[1.2rem] -rotate-90 scale-100' />
						<span className='sr-only'>Open Menu</span>
					</Button>
				</SheetTrigger>
				<SheetContent side='left' className='p-4'>
					<ul className='space-y-4 py-4'>
						{filterLinkItems().map(({ href, name, Icon }) => (
							<li key={name}>
								<NavigationLink
									href={href}
									currentPath={pathname}
									onClick={() => setIsOpen(false)}>
									<Icon size={20} />
									<span>{name}</span>
								</NavigationLink>
							</li>
						))}
					</ul>
				</SheetContent>
			</Sheet>
		</>
	)
}

export const Navigation = () => {
	const pathname = usePathname()
	const { status } = useSession()
	const { isCollapse, isOpen, setIsCollapse, setIsOpen } = useNavigationState()

	const filterLinkItems = () => {
		if (status === 'authenticated') {
			return menuLinkItems.filter(({ visible }) => visible !== 'no-auth')
		}

		return menuLinkItems.filter(({ visible }) => visible !== 'auth')
	}

	return (
		<>
			<nav
				className={cn(
					'sticky top-0 hidden min-h-screen w-48 flex-col items-center self-start px-4 py-8 sm:flex',
					'translate-y-0 transition-[width,transform] duration-300 motion-reduce:transition-none',
					{ 'w-[5.25rem] -translate-y-[14.75rem] delay-150': isCollapse },
				)}>
				<div
					className={cn(
						'w-full overflow-hidden pb-4 flex flex-col items-center',
						{
							'space-y-2': !isCollapse,
						},
					)}>
					<div
						className={cn(
							'h-36 w-36 overflow-hidden rounded-3xl',
							'translate-x-0 transition-transform duration-300 motion-reduce:transition-none',
							{ '-translate-x-48': isCollapse, 'delay-150': !isCollapse },
						)}>
						<ImageBlur
							blurDataURL='/assets/speakfest.png'
							src='/assets/speakfest.png'
							width={246}
							height={246}
							alt='Avatar'
						/>
					</div>
					<Heading
						variant='h2'
						className={cn(
							'translate-x-0 transition-transform duration-300 motion-reduce:transition-none',
							{ '-translate-x-48': isCollapse, 'delay-150': !isCollapse },
						)}>
						SpeakFest
					</Heading>
					{/* <ThemeToggle isCollapse={isCollapse} /> */}
				</div>
				{status === 'loading' ? (
					<ul className='w-full space-y-4 border-y py-4'>
						<NavigationLinkSkeleton />
					</ul>
				) : (
					<ul className='w-full space-y-4 border-y py-4'>
						{filterLinkItems().map(({ href, name, Icon }) => (
							<li key={name}>
								<NavigationLink
									href={href}
									currentPath={pathname}
									isCollapse={isCollapse}>
									<Icon size={20} />
									<span className={cn({ 'sr-only': isCollapse })}>{name}</span>
								</NavigationLink>
							</li>
						))}
					</ul>
				)}

				<Button
					size='icon'
					variant='outline'
					className='mt-4 w-full'
					onClick={() => setIsCollapse((prev) => !prev)}>
					{isCollapse ? (
						<ChevronsRight size={20} />
					) : (
						<ChevronsLeft size={20} />
					)}
					<span className='sr-only'>Collapse sidebar</span>
				</Button>
			</nav>

			<MobileNavigation
				isOpen={isOpen}
				pathname={pathname}
				setIsOpen={setIsOpen}
				filterLinkItems={filterLinkItems}
			/>
		</>
	)
}
