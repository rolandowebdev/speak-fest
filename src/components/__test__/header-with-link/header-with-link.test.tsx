/**
 * Scenario test for HeaderWithLink component :
 * - should navigate to home page when clicked back to home button
 */

import * as React from 'react'
import { describe, it } from '@jest/globals'
import { render, fireEvent, screen } from '@testing-library/react'
import { HeaderWithLink } from '@/components/custom'
import { useRouter } from 'next/navigation'
import { Home } from 'lucide-react'

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}))

describe('HeaderWithLink Component', () => {
  const pushMock = jest.fn()

  beforeEach(() => {
    ;(useRouter as jest.Mock).mockReturnValue({
      push: pushMock,
    })
  })

  it('should navigate to home page when clicked back to home button', () => {
    const view = render(<HeaderWithLink icon={<Home />} title="Title" />)
    const backButton = screen.getByTestId('back-to-home')

    fireEvent.click(backButton)

    expect(pushMock).toHaveBeenCalledWith('/')
    expect(pushMock).toHaveBeenCalledTimes(1)
    expect(view).toMatchSnapshot()
  })
})
