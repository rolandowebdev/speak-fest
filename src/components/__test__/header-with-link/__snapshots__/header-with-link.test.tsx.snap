// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderWithLink Component should navigate to home page when clicked back to home button 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="space-y-4 border-b pb-4"
      >
        <button
          class="justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 link-variant-test underline-offset-4 hover:underline default-size-test h-10 py-2 flex items-center gap-1 px-0 text-lg text-primary"
          data-testid="back-to-home"
        >
          <svg
            class="lucide lucide-undo2 "
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 14 4 9l5-5"
            />
            <path
              d="M4 9h10.5a5.5 5.5 0 0 1 5.5 5.5v0a5.5 5.5 0 0 1-5.5 5.5H11"
            />
          </svg>
          Back to home
        </button>
        <h1
          class="scroll-m-12 font-bold text-4xl flex flex-wrap items-center gap-2"
        >
          <svg
            class="lucide lucide-home "
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m3 9 9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
            />
            <polyline
              points="9 22 9 12 15 12 15 22"
            />
          </svg>
           
          Title
        </h1>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="space-y-4 border-b pb-4"
    >
      <button
        class="justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 link-variant-test underline-offset-4 hover:underline default-size-test h-10 py-2 flex items-center gap-1 px-0 text-lg text-primary"
        data-testid="back-to-home"
      >
        <svg
          class="lucide lucide-undo2 "
          fill="none"
          height="18"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 14 4 9l5-5"
          />
          <path
            d="M4 9h10.5a5.5 5.5 0 0 1 5.5 5.5v0a5.5 5.5 0 0 1-5.5 5.5H11"
          />
        </svg>
        Back to home
      </button>
      <h1
        class="scroll-m-12 font-bold text-4xl flex flex-wrap items-center gap-2"
      >
        <svg
          class="lucide lucide-home "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 9 9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
          />
          <polyline
            points="9 22 9 12 15 12 15 22"
          />
        </svg>
         
        Title
      </h1>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
